File: Behavioural-Patterns/Memento/C#/Memento/Implementations/Canvas.cs
public class Canvas {
    private List<Shape> shapes = new List<Shape>();

    public void addShape(Shape shape){
        shapes.Add(shape);
        Console.WriteLine("Added shape: " + shape);
    }

    public CanvasState createMemento(){
        List<Shape> savedShapes = shapes.Select(shape => new Shape(shape)).ToList();
        return new CanvasState(savedShapes);
    }

    public void restore(CanvasState memento){
        this.shapes = memento.getSavedShapes();
        Console.WriteLine("Canvas has been restored.");
    }

    public void printShapes() {
        if (shapes.Count == 0) {
            Console.WriteLine("Canvas is empty.");
        } else {
            Console.WriteLine("Current Canvas Shapes:");
            shapes.ForEach(Console.WriteLine);
        }
        Console.WriteLine("-------------------------");
    }

    public class CanvasState{
        internal readonly List<Shape> savedShapes;

        internal CanvasState(List<Shape> shapes){
            this.savedShapes = shapes;
        }

        internal List<Shape> getSavedShapes(){
            return this.savedShapes;
        }
    }
}

File: Behavioural-Patterns/Memento/Java/Implementations/Canvas.java
package Implementations;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Canvas {
    private List<Shape> shapes = new ArrayList<>();

    public void addShape(Shape shape){
        shapes.add(shape);
        System.out.println("Added shape: " + shape);
    }

    public CanvasState createMemento(){
        List<Shape> savedShapes = shapes.stream().map(Shape::new).collect(Collectors.toList());
        return new CanvasState(savedShapes);
    }

    public void restore(CanvasState memento){
        this.shapes = memento.getSavedShapes();
        System.out.println("Canvas has been restored.");
    }

    public void printShapes() {
        if (shapes.isEmpty()) {
            System.out.println("Canvas is empty.");
        } else {
            System.out.println("Current Canvas Shapes:");
            shapes.forEach(System.out::println);
        }
        System.out.println("-------------------------");
    }

    public static class CanvasState{
        private final List<Shape> savedShapes;

        private CanvasState(List<Shape> shapes){
            this.savedShapes = shapes;
        }

        private List<Shape> getSavedShapes(){
            return this.savedShapes;
        }
    }
}
